### PR TITLE
Clean up chat input components

### PR DESCRIPTION
**Pull request by akalin-keybase**
Converted a bunch of recompose stuff and a HOC to two wrapper components.
Now we have:
   Input, with props InputProps, which wraps
     MentionInput, with props MentionInputProps, which wraps
       PlatformInput, with props PlatformInputProps.

The logic was way confusing before, and everything was untyped, so now
it should be easier to modify.

This is in preparation for using an unconnected TextInput component, as well
as de-platforming the MentionInput component.

### CODE REVIEW

File: shared/chat/conversation/input-area/normal/platform-input.native.js [hunk 1]
```diff
@@ -80,6 +80,7 @@ class ConversationInput extends Component<Props> {
             hideUnderline={true}
             hintText={hintText}
             multiline={true}
+            onBlur={this.props.onBlur}
```
Comment (akalin-keybase) [inline shared/chat/conversation/input-area/normal/platform-input.native.js]: We were dropping this!

File: shared/chat/conversation/input-area/normal/platform-input.desktop.js [hunk 2]
```diff
@@ -347,24 +395,4 @@ const styleFooter = platformStyles({
   },
 })
 
-export default compose(
```
Comment (akalin-keybase) [inline shared/chat/conversation/input-area/normal/platform-input.desktop.js]: Just moved this into the component

File: shared/chat/conversation/input-area/normal/container.js [hunk 3]
```diff
@@ -134,119 +140,18 @@ const mergeProps = (stateProps, dispatchProps, ownProps: OwnProps) => ({
...
-  _quotingMessage: ?Types.Message,
-  _editingMessage: ?Types.Message,
-  setText: (string, skipUnsentSaving?: boolean) => void,
-  injectedInput: string,
-  inputMoveToEnd: () => void,
-}
 
-export default compose(
-  connect(mapStateToProps, mapDispatchToProps, mergeProps),
-  withStateHandlers(
```
Comment (akalin-keybase) [inline shared/chat/conversation/input-area/normal/container.js]: moved most of this logic into index.js

File: shared/chat/conversation/input-area/normal/container.js [hunk 4]
```diff
@@ -134,119 +140,18 @@ const mergeProps = (stateProps, dispatchProps, ownProps: OwnProps) => ({
...
-    onSubmit: (text: string) => {
-      props._onSubmit(text)
-      props.setText('')
-    },
-  })),
-  withHandlers(props => {
-    let input
-    return {
-      _inputSetRef: props => i => (input = i),
-      _onKeyDown: props => (e: SyntheticKeyboardEvent<>) => {
```
Comment (akalin-keybase) [inline shared/chat/conversation/input-area/normal/container.js]: moved this logic into `platform-input.desktop.js`

File: shared/chat/conversation/input-area/normal/index.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+// Standalone throttled function to ensure we never accidentally recreate it and break the throttling
+const throttled = throttle((f, param) => f(param), 1000)
+
+class Input extends React.Component<InputProps, State> {
+  _input: ?TextInput
+
+  constructor(props: InputProps) {
+    super(props)
+    this.state = {
```
Comment (akalin-keybase) [inline shared/chat/conversation/input-area/normal/index.js]: this state will go away when we use unconnected TextInput

File: shared/chat/conversation/input-area/normal/mention-input.desktop.js [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+  downArrowCounter: number,
+  pickSelectedCounter: number,
+  mentionFilter: string,
+  channelMentionFilter: string,
+  mentionPopupOpen: boolean,
+  channelMentionPopupOpen: boolean,
+}
+
+class MentionInput extends React.Component<MentionInputProps, MentionState> {
+  state: MentionState = {
```
Comment (akalin-keybase) [inline shared/chat/conversation/input-area/normal/mention-input.desktop.js]: didn't change most of the logic, but converted from a HOC to a wrapper component, so lotta whitespace diffs

File: shared/chat/conversation/input-area/normal/mention-input.native.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
+// @flow
```
Comment (akalin-keybase) [inline shared/chat/conversation/input-area/normal/mention-input.native.js]: same situation as corresponding desktop file

**Comment by akalin-keybase** [comment]
This is ready. r? @keybase/react-hackers

**Comment by chrisnojima** [comment]
code looks fine but i need someone to actually run it and test the various flows. things to check
- [ ] editing works
- [ ] up arrow editing works
- [ ] typing and not sending and switching to a different convo and switching back retains the text
- [ ] at mentions work
- [ ] channel mentions work

**Comment by akalin-keybase** [comment]
Couple more items that needs verification:

[ ] Quoting a message works
[ ] 'Reply Privately' works

**Comment by buoyad** [comment]
Tests:

1. Editing works, :sheep: on mobile the yellow marker isn't full height: image
2. Up arrow editing is fine (tested on mac & ubuntu)
3. Pending text is retained when switching around conversations
4. @-mentions work, but 'channel' shows up in the list when in ad-hoc conversations (just checked and that's how it already works)
5. #-mentions are good, but the list pops up in small teams (just checked and that's how it already works)
6. Quoting doesn't always work on mobile
    * I had it work once, but most times (maybe every time after the first) it doesn't show in the input
    * When I back out of the conversation and back in, the quoted text appears correctly
7. Reply privately works. The quote shows up most times on mobile, but there's a lag before it's filled in

**Comment by akalin-keybase** [comment]
Thanks for testing! Some notes for the results:

1. Looks like the yellow box on mobile not being full height was already there, so I'll punt, but added a TODO.
6. Can't repro this yet :/
7. Yeah, I noticed the lag. I traced through and nothing is obviously broken, but may be because of it being controlled -- seems to show up faster on  https://github.com/keybase/client/pull/11622

Also found a bug where after quoting on mobile and editing the quote, going to another channel and back restores the full quote. That bug was already on master, so will fix separately, and added a TODO.

**Comment by akalin-keybase** [comment]
Repro'd #2, but it's also on master so I'll punt. Interpreting @chrisnojima's approval on https://github.com/keybase/client/pull/11622 to include this one, and merging.

**Pull request merged**